### ISSUE TITLE
[question] conan inspect

### ISSUE THREAD

**Issue by db4** [conan-io/conan#14191]
### What is your question?

`conan inspect` in Conan V2 doesn't generate boolean values in JSON output. Compare inspection result of zlib for V1 and V2:
V1:
```json
{
    "name": "zlib",
    "version": null,
    "url": "https://github.com/conan-io/conan-center-index",
    "homepage": "https://zlib.net",
    "license": "Zlib",
    "author": null,
    "description": "A Massively Spiffy Yet Delicately Unobtrusive Compression Library (Also Free, Not to Mention Unencumbered by Patents)",
    "topics": [
        "zlib",
        "compression"
    ],
    "generators": [
        "txt"
    ],
    "exports": null,
    "exports_sources": null,
    "short_paths": false,
    "apply_env": true,
    "build_policy": null,
    "revision_mode": "hash",
    "settings": [
        "os",
        "arch",
        "compiler",
        "build_type"
    ],
    "options": {
        "shared": [
            true,
            false
        ],
        "fPIC": [
            true,
            false
        ]
    },
    "default_options": {
        "shared": false,
        "fPIC": true
    },
    "deprecated": null
}
```
V2:
```json
{
    "name": "zlib",
    "user": null,
    "channel": null,
    "url": "https://github.com/conan-io/conan-center-index",
    "license": "Zlib",
    "author": null,
    "description": "A Massively Spiffy Yet Delicately Unobtrusive Compression Library (Also Free, Not to Mention Unencumbered by Patents)",
    "homepage": "https://zlib.net",
    "build_policy": null,
    "upload_policy": null,
    "revision_mode": "hash",
    "provides": null,
    "deprecated": null,
    "win_bash": null,
    "win_bash_run": null,
    "default_options": {
        "shared": false,
        "fPIC": true
    },
    "options_description": null,
    "version": null,
    "topics": [
        "zlib",
        "compression"
    ],
    "package_type": "library",
    "settings": [
        "os",
        "arch",
        "compiler",
        "build_type"
    ],
    "options": {
        "fPIC": "True",
        "shared": "False"
    },
    "options_definitions": {
        "shared": [
            "True",
            "False"
        ],
        "fPIC": [
            "True",
            "False"
        ]
    },
    "generators": [],
    "requires": [],
    "source_folder": null,
    "build_folder": null,
    "generators_folder": null,
    "package_folder": null,
    "label": ""
}
```
Conan V1 correctly generates boolean values, but V2 produces strings instead. Also `conan inspect` in Conan V2 does not accept references. That's quite inconvenient.

BTW, can I fetch recipe options via Python API? I don't need the full graph, just possible options (for my build scripts)

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by AbrilRBS** [comment]
Hi @db4 thanks a lot for taking the time to report the issue :)

I see the issue and how it affects automatic flows, it's totally reasonable to expect True/False to be the proper json boolean value.

Let me take a look at how the current code handles this, and will let you know once the fix is ready/a decision not to change anything is made 😬 


> can I fetch recipe options via Python API?

Yes! Using `conan_api.local.inspect(path_to_conanfile)` will give you the same json you're seeing as the output of `conan inspect` (Plus some extra fields that are removed from the command as they are not initialized), and from there you can fetch your information as needed

**Comment by AbrilRBS** [comment]
Adding this as a look-into to discuss with the team next week, as after checking the code, it seems that they being strings is the expected behaviour from conan, and chaning it just to display proper types might be an unwanted tradeoff

**Comment by AbrilRBS** [comment]
Hi @db4, thanks for your patience. We'll be looking into serializing the options and a few other things into proper boolean values when you would expect them to.

I also see that you mention that:

> Also conan inspect in Conan V2 does not accept references. That's quite inconvenient.

What is the use case here? What would you like to be able to do here? The current reasoning for not allowing references to be inspected is that it would need to bring the recipe to the local cache, and then using that library as a dependency would lead to unexpected behaviour where it would use it that version even if no remotes were passed to conan when installing/creating the package.

Maybe a tradeoff solution is to only allow inspect for references that are already in the cache? Would that work for you?

**Comment by db4** [comment]
> > Also conan inspect in Conan V2 does not accept references. That's quite inconvenient.

> What is the use case here? What would you like to be able to do here?

In my existing Conan 1.x CI scripts that I'm porting to Conan 2.x, I export a bunch of recipes to the local cache, analyze dependencies between them, sort them topologically, check their options, and build missing binary packages for the computed set of options. If I cannot get options by reference, I should store `conanfile.py` path somewhere. Not a big deal, actually I have already done that.

**Comment by AbrilRBS** [comment]
Thanks for the extra insight :)

 * Why are you topo-sorting the dependencies? Is `conan build-order` not enough for your use-case?  We would love to know if that's the case to better understand its shortcomings.
 * If you are exporting the recipes, you already have the `conanfile.py`'s path, for which you could call `inspect` at that point. Is that not possible? It would avoid needing the reference later.
 * Last, I want to touch on a point that scares me a bit: You say you are building for the computed set of options! Do you have any process in place to avoid exponential explosion of your build matrix? It sounds a bit risky the way you put it, so maybe I misunderstood! Or are you only restricting yourself to a set of "known good" options? That sounds safer, but would mean there would be no need to actually compute anything, just pre-compute your wanted profiles and only build those.

**Comment by db4** [comment]
Hi @RubenRBS,
It needs a longer explanation, but if you are asking for it...

I had a CCI-like recipe repository. Since CCI CI scripts were closed I had to invent my own ones from scratch. CCI scripts allowed to commit only one recipe at a time and assumed that everything else is up-to-date. For me, it was a too strong limitation (sometimes it was convenient to commit a group of related recipes checking that everything went well), and even then it didn't work well - see numerous CCI issues regarding missing binary packages including my own opencv PR.

So I chose another approach for CI: export **everything** to the local cache (all recipes/versions) first and then find packages that have to be built. Roughly I do the following:

- scan the file system and `conan export` all recipes found, also fill a Python dictionary with the found references data, that is used/updated through the next steps
- for every reference:
  -  determine which set of options is needed (intersection of `conan inspect` values and global/per-recipe `buildopts.py` values). E.g. if the recipe has `shared` option, both `shared=False` and `shared=True` binaries are needed. Combine the computed options with the predefined set of settings.
  - loop through computed options/settings set and check if the binary package is missing (`conan info`)
- topologically sort the missing binaries. That cannot be done by `conan build-order`, since there is no common graph for everything that needs to be built
- build the missing binaries (`conan create`)
- if everything is built successfully upload the binaries to a private Conan repo

You may ask about performance: isn't it require too much time to export/search everything? Well, I have about 1000 unique binaries to check it takes about ten minutes. But I also have an optimization that saves sha256 of the last successful commit and if it's present only changed recipes (taken by `git diff ... --name-status`) are searched. That speeds up the search process dramatically.

**Comment by memsharded** [comment]
Thanks for the explanations.

I have in my branch https://github.com/memsharded/conan-center-index/blob/mycustom/mybuild.py a very simple approach to create my own binaries from a conan-center-index fork:

- No checking if binaries exist with ``conan info`` is necessary, ``conan create --build=missing`` will do it.
- No topological order is needed, I just run sequentially the ``conan create --build=missing``. But if a topological order is necessary, Conan 2.0 can do it with ``build-order`` of every one, then ``conan graph build-order-merge`` into a single build-order, it is built-in
- The fact that a recipe has shared option or not is not very relevant to me, I can just define a ``*:shared=True`` profile, that will apply it for the recipes that have the option defined, and ignore the others. Still, I know the combinations that I want to build a-priori, it doesn't depend on whatever the recipes define (I'd say this is what you already have in your ``buildopts.py``, just you don't need to inspect the recipes to check for the options, but can apply them with patterns)

I have found this approach to be super simple, effective and efficient enough. I understand you might have other requisites, but I thought it was relevant to share it.

**Comment by db4** [comment]
@memsharded,
Thanks for your suggestions. I added `conan info` step initially because Conan package tools I was using back then always ran  test_package on build=outdated (see https://github.com/conan-io/conan-package-tools/issues/434). I'm in the process of getting rid of CPT even for Conan 1.x, so maybe `conan create --build=missing` will really be enough.

As for
> I can just define a `*:shared=True` profile, that will apply it for the recipes that have the option defined, and ignore the others

It means that you create all packages twice: with `shared` and `non_shared` profiles. This is unneeded work for packages that don't have `shared` option. Some of my recipes have other internal common options. If you have 3 of them, that's 8 profiles in total. And you should create those profiles manually. For me examining available options during runtime and generating the necessary builds dynamically is more convenient.

**Comment by AbrilRBS** [comment]
> It means that you create all packages twice: with shared and non_shared profiles. This is unneeded work for packages that don't have shared option

Conan will only create the package the first time around. Only the options defined in a recipe are taking into account when computing the package id, so calling `conan create` with unknown options just ignores them.

After the first creation, that package will be reused as long as the computed id matches :)

**Comment by AbrilRBS** [comment]
Closing as solved, please feel free to reopen/create new issues if more help is needed, thanks!
